### DISCUSSION TITLE
Use type predicates for `isLocale` and `isNamespace` functions

### DISCUSSION THREAD

**Discussion by mateuszaliyev** [ivanhofer/typesafe-i18n#475] [DX improvements]
I think recently added `isLocale` and `isNamespace` functions would be perfect use cases for [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates). Thanks to this feature TypeScript would be able to narrow down the type of passed locale/namespace parameters.

Current implementation:

![image](https://user-images.githubusercontent.com/82393815/195699080-7669f8cd-a5d6-4a91-bba3-1c2ffb875642.png)

With type predicates:

![image](https://user-images.githubusercontent.com/82393815/195699166-162d6959-75f8-4ccc-8958-373f24136ad4.png)

Adding `locale is Locales` and `namespace is Namespaces` as a return type of `isLocale` and `isNamespace` functions is all that's needed. Note that it should also work with JSDoc `@returns` annotation (`@returns {locale is Locale}`).

I tried implementing this feature myself by following the [`CONTRIBUTING.md`](https://github.com/ivanhofer/typesafe-i18n/blob/main/CONTRIBUTING.md) file but I couldn't figure out the testing part of generator files (regardless of multiple `pnpm test` and `pnpm test:update-generated-files` calls, it would not generate any files) so I'm creating a discussion instead. Just in case it happens to be useful, those are the changes I made:

<details>
<summary><code>packages/generator/src/files/generate-util.mts</code></summary>

```ts
`
${jsDocFunction('locale is Locales', {
  type: 'string',
  name: 'locale',
})}
export const isLocale = (locale${type('string')})${typePredicate('locale', 'Locales')} => locales.includes(${jsDocType(
  'Locales',
  `locale${typeCast('Locales')}`,
)})
${usesNamespaces ? `
${jsDocFunction('namespace is Namespaces', {
 type: 'string',
 name: 'namespace',
})}
export const isNamespace = (namespace${type('string')})${typePredicate('namespace', 'Namespaces')} => namespaces.includes(${jsDocType(
  'Namespaces',
  `namespace${typeCast('Namespaces')}`,
)})` : ''}
`
```
</details>
<details>
<summary><code>packages/generator/src/output-handler.mts</code></summary>

```ts
export const configureOutputHandler = (config: GeneratorConfigWithDefaultValues, version: TypescriptVersion): void => {

  /* ... */
  
  typePredicate = (param, type) => (shouldGenerateJsDoc ? '' : ` ${param} is ${type}`)
  
  /* ... */
  
}

/* ... */

export let typePredicate: (param: string, type: string) => string
```
</details>

**Comment by ivanhofer** [comment]
Thanks @mateuszaliyev! Indeed this would be a great feature, I didn't think of when I have implemented the function.

Can you please share what didn't work for you?

And thanks for your detailed examples!

**Reply by mateuszaliyev** [reply]
Glad I could help!

I tried to follow instructions inside of `CONTRIBUTING.md` file. I forked the repo, cloned it, ran `pnpm test` and `pnpm test:update-generated-files`. I'm probably understanding it wrong though, but I expected some files like `packages\generator\test\generated\adapter-react\util.expected.ts` to reflect my changes, [just like here](https://github.com/ivanhofer/typesafe-i18n/commit/cf6f2d914ae6bbe082d52f9b8c12d9939f5b2834). Nothing was generated, so I've decided to create a discussion instead.

**Reply by ivanhofer** [reply]
What operating system are you using? Which node version?

**Reply by mateuszaliyev** [reply]
Windows 11 Home 21H2
Node v16.15.1

**Reply by ivanhofer** [reply]
Ok, I'll try to take a look this week and see if I can reproduce the issue

**Comment by ivanhofer** [comment]
I have made the changes and released version `5.15.0` right now
